core/src/avm2/array.rs

```diff
@@ -33,49 +97,191 @@ impl<'gc> ArrayStorage<'gc> {
+                if storage.len() < (item + 1) {
+                    if *occupied_count < (item / 4) && MIN_SPARSE_LENGTH < item
+                        || item >= MAX_DENSE_LENGTH
+                    {
+                        self.convert_to_sparse();
+                        if let ArrayStorage::Sparse { storage, length } = self {
+                            *length = item + 1;
+                            storage.insert(item, value);
+                        }
+                    } else {
+                        storage.resize(item + 1, None);
+                        *storage.get_mut(item).unwrap() = Some(value);
+                        let new_holes = (item + 1) - storage.len();
+                        *occupied_count = (*occupied_count + new_holes) + 1;
+                    }
+                } else {
+                    if storage[item].is_none() {
+                        *occupied_count += 1;
+                    }
+                    *storage.get_mut(item).unwrap() = Some(value);
+                }
+            }
+            ArrayStorage::Sparse { storage, length } => {
+                storage.insert(item, value);
+                if item >= *length {
+                    *length = item + 1;
+                }
+            }
         }
-
-        *self.storage.get_mut(item).unwrap() = Some(value)
     }
 
     /// Delete an array storage slot, leaving a hole.
     pub fn delete(&mut self, item: usize) {
-        if let Some(i) = self.storage.get_mut(item) {
-            *i = None;
+        match self {
+            ArrayStorage::Dense {
+                storage,
+                occupied_count,
+            } => {
+                if let Some(i) = storage.get_mut(item) {
+                    *i = None;
+                    if *occupied_count != 0 {
+                        *occupied_count -= 1;
+                        self.maybe_convert_to_sparse();
+                    }
+                }
+            }
+            ArrayStorage::Sparse { storage, .. } => {
+                storage.remove(&item);
+                self.maybe_convert_to_dense();
+            }
         }
     }
 
     /// Get the length of the array.
     pub fn length(&self) -> usize {
-        self.storage.len()
+        match &self {
+            ArrayStorage::Dense { storage, .. } => storage.len(),
+            ArrayStorage::Sparse { length, .. } => *length,
+        }
     }
 
     /// Set the length of the array.
     pub fn set_length(&mut self, size: usize) {
-        self.storage.resize(size, None)
+        match self {
+            ArrayStorage::Dense {
+                storage,
+                occupied_count,
+            } => {
+                if size < MAX_DENSE_LENGTH {
+                    let mut num_of_new_holes = 0;
+                    if size > storage.len() {
+                        num_of_new_holes = size - storage.len();
+                    }
+                    if *occupied_count + num_of_new_holes < (size / 4)
```

Comment:
is this math correct? Not sure why it's adding holes to occupied.
Also, how about extracting the repeated logic to a function? As in, the
```
like, new_size >= MIN_SPARSE_LENGTH && new_used < new_size / 4
```